aten/src/ATen/core/type.cpp

```diff
@@ -70,6 +71,43 @@ AnyTypePtr AnyType::get() {
   return value;
 }
 
+template <typename T>
+static bool test_primitive(c10::optional<T> e, T a) {
+  return !e.has_value() || e.value() == a;
+}
+
+static bool test_varying_shape(const VaryingShape& e, at::IntArrayRef a) {
+  if (!e.size().has_value()) {
+    return true;
+  }
+
+  if (e.size().value() != a.size()) {
+    return false;
+  }
+
+  auto ndim = a.size();
+  for (size_t i = 0; i < ndim; i++) {
+    if (!test_primitive(e[i], a[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool TensorType::isCompatibleWith(at::Tensor& t) const {
```

Comment:
nit: isCompatibleWithInCurrentExecutionContext

The second part is important since it would otherwise be surprising that we are looking at GradMode in this function.